mono/mini/debugger-agent.c

```diff
@@ -4664,6 +4664,42 @@ ss_update (SingleStepReq *req, MonoJitInfo *ji, SeqPoint *sp, DebuggerTlsData *t
 	MonoDebugSourceLocation *loc = NULL;
 	gboolean hit = TRUE;
 
+	if ((ss_req->filter & STEP_FILTER_STATIC_CTOR)) {
+		mono_thread_state_init_from_monoctx (&tls->context, ctx);
+		compute_frame_info (tls->thread, tls);
+
+		gboolean ret = FALSE;
+		gboolean method_in_stack = FALSE;
+
+		for (int i=0; i < tls->frame_count; i++) {
+			MonoMethod *external_method = tls->frames [i]->method;
+				if (method == external_method)
+					method_in_stack = TRUE;
+
+				if (!ret) {
+					ret = (external_method->flags & METHOD_ATTRIBUTE_SPECIAL_NAME);
+					ret = ret && !strcmp (external_method->name, ".cctor");
+					ret = ret && (external_method != req->start_method);
+				}
+		}
+
+		if (!method_in_stack) {
+			DEBUG_PRINTF (1, "[%p] The instruction pointer of the currently executing method(%s) is not on the recorded stack. This is likely due to a runtime bug. The %d frames are as follow: \n", (gpointer)(gsize)mono_native_thread_id_get (), mono_method_full_name (method, TRUE), tls->frame_count);
+
+			for (int i=0; i < tls->frame_count; i++)
+				DEBUG_PRINTF (1, "\t [%p] Frame (%d / %d): %s\n", (gpointer)(gsize)mono_native_thread_id_get (), i, tls->frame_count, mono_method_full_name (tls->frames [i]->method, TRUE));
+		}
+		g_assert (method_in_stack);
+
+
+		tls->context.valid = FALSE;
+		tls->async_state.valid = FALSE;
+		invalidate_frames (tls);
+
```

Comment:
Why do we have to invalidate frames here ?

Reply:
If we don't, there are two issues. The first is that on some code paths, we will leak memory (according to Instruments). The second is that if we don't invalidate the frames, we won't recompute them at other, later points. This non-recomputing leads to tests failing.

Reply:
The frames are supposed to be invalidated when the thread is restarted.